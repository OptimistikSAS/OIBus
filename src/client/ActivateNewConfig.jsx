import React from 'react'
import { Label, Button } from 'reactstrap'
import { ReactGhLikeDiff } from 'react-gh-like-diff'
import stringify from 'json-stable-stringify'
import 'react-gh-like-diff/lib/diff2html.min.css'
import apis from './services/apis'


const ActivateNewConfig = () => {
  const [configActiveJson, setConfigActiveJson] = React.useState(null)
  const [configJson, setConfigJson] = React.useState(null)

  /**
   * Acquire the Active configuration
   * @returns {void}
   */
  React.useEffect(() => {
    apis.getActiveConfig().then(({ config }) => {
      setConfigActiveJson(config)
    })
  }, [])

  /**
   * Acquire the Current Modified configuration
   * @returns {void}
   */
  React.useEffect(() => {
    apis.getConfig().then(({ config }) => {
      setConfigJson(config)
    })
  }, [])

  /**
   * Check if modified is changed compared to active
   * @param {string} active The current active config JSON string
   * @param {string} modified The current modified config JSON string
   * @returns {boolean} compare result
   */
  const compareActiveWithModified = (active, modified) => (active && modified && (active !== modified))

  /**
   * Activate new configuration
   * @returns {void}
   */
  const handleActivate = async () => {
    try {
      await apis.updateActiveConfig()
      setConfigActiveJson(configJson)
    } catch (error) {
      console.error(error)
    }
  }

  /**
   * Reset modified configuration
   * @returns {void}
   */
  const handleDecline = async () => {
    try {
      await apis.resetModifiedConfig()
      setConfigJson(configActiveJson)
    } catch (error) {
      console.error(error)
    }
  }

  // json-stable-stringify is used instead of JSON.stringify to have consistent result in alphabetical order
  const activeString = stringify(configActiveJson, { space: '  ' })
  const modifiedString = stringify(configJson, { space: '  ' })
  const isModified = compareActiveWithModified(activeString, modifiedString)

  return (
    <>
      {isModified
        ? (
          <>
            <ReactGhLikeDiff
              options={{
                originalFileName: 'Configuration',
                updatedFileName: 'Configuration',
              }}
              past={activeString}
              current={modifiedString}
            />
            <Button color="primary" onClick={() => handleActivate()}>
              Activate
            </Button>
            <Button color="primary" onClick={() => handleDecline()}>
              Decline
            </Button>
          </>
        )
        : <Label>No modifications on configuration</Label>
      }
    </>
  )
}

export default ActivateNewConfig
