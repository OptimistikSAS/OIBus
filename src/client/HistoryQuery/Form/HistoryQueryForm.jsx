import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Button, Col, Container, Form, Label, Row } from 'reactstrap'
import { useNavigate } from 'react-router-dom'
import { FaArrowLeft, FaArrowRight, FaPauseCircle, FaPlayCircle } from 'react-icons/fa'
import { OIbCheckBox, OIbInteger, OIbText, OIbTextArea, OIbTitle } from '../../components/OIbForm'
import OIbDate from '../../components/OIbForm/OIbDate.jsx'
import { ConfigContext } from '../../context/ConfigContext.jsx'
import PointsSection from './PointsSection.jsx'
import apis from '../../services/apis'
import utils from '../../helpers/utils'
import { AlertContext } from '../../context/AlertContext.jsx'
import ConnectorButton from '../ConnectorButton.jsx'

const HistoryQueryForm = ({ query }) => {
  const [queryToUpdate, setQueryToUpdate] = useState(query)
  const [progressStatus, setProgressStatus] = useState({ status: query.status })

  const { name, paused } = queryToUpdate
  const { newConfig } = React.useContext(ConfigContext)
  const [errors, setErrors] = useState({})
  const dataSource = newConfig?.south?.dataSources.find(
    (southHandler) => southHandler.id === queryToUpdate.southId,
  )
  const application = newConfig?.north?.applications.find(
    (northHandler) => northHandler.id === queryToUpdate.northId,
  )

  const { setAlert } = React.useContext(AlertContext)
  const navigate = useNavigate()

  React.useEffect(() => {
    let source
    if (progressStatus.status !== 'finished' && progressStatus.status !== 'pending') {
      source = new EventSource(`/history/${queryToUpdate.id}/sse`)
      source.onerror = (error) => {
        console.error(error)
      }

      source.onmessage = (event) => {
        if (event && event.data) {
          const myData = JSON.parse(event.data)
          setProgressStatus({
            currentTime: myData.currentTime,
            progress: myData.progress,
            scanGroup: myData.scanGroup,
            status: myData.status,
          })
        }
      }
    }

    return () => {
      source?.close()
    }
  }, [queryToUpdate])

  const handlePause = async () => {
    await apis.pauseHistoryQuery(queryToUpdate.id, { paused: !paused })
    setQueryToUpdate({ ...queryToUpdate, paused: !queryToUpdate.paused })
  }

  const handleEnabled = async () => {
    await apis.enableHistoryQuery(queryToUpdate.id, { enabled: !queryToUpdate.enabled })
    setQueryToUpdate({ ...queryToUpdate, enabled: !queryToUpdate.enabled })
  }

  const handleUpdateHistoryQuery = async () => {
    if (Object.values(errors).length === 0) {
      setAlert(null)
      await apis.updateHistoryQuery(queryToUpdate.id, { ...queryToUpdate })
    } else {
      setAlert({
        text: 'You have unresolved errors in configuration! Please fix before updating the settings.',
        type: 'danger',
      })
    }
  }

  const onChange = (propertyName, value) => {
    setQueryToUpdate({ ...queryToUpdate, [propertyName]: value })
  }

  const onMaxReadIntervalChange = (_, value) => {
    setQueryToUpdate({ ...queryToUpdate, settings: { ...queryToUpdate.settings, maxReadInterval: value } })
  }

  const onReadIntervalDelayChange = (_, value) => {
    setQueryToUpdate({ ...queryToUpdate, settings: { ...queryToUpdate.settings, readIntervalDelay: value } })
  }

  const handleAddQuery = (_, value) => {
    setQueryToUpdate({ ...queryToUpdate, settings: { ...queryToUpdate.settings, query: value } })
  }

  const handleAddPoint = (attributes) => {
    const newPointAttributes = {}
    attributes.forEach((attribute) => {
      newPointAttributes[attribute] = ''
    })
    setQueryToUpdate(
      {
        ...queryToUpdate,
        settings: {
          ...queryToUpdate.settings,
          points: [...queryToUpdate.settings.points, { ...newPointAttributes }],
        },
      },
    )
  }

  const handleChangePoint = (attributeDescription, value, validity) => {
    if (validity) {
      setErrors((oldErrors) => ({ ...oldErrors, [attributeDescription]: validity }))
    } else {
      const copyErrorObject = errors
      delete copyErrorObject[attributeDescription]
      setErrors(copyErrorObject)
    }

    const attributeName = attributeDescription.split('.')[2]
    const pointIndex = Number(attributeDescription.split('.')[1])
    setQueryToUpdate((oldQuery) => ({
      ...oldQuery,
      settings: {
        ...oldQuery.settings,
        points: oldQuery.settings.points.map((point, index) => {
          if (pointIndex === index) {
            return { ...point, [attributeName]: value }
          }
          return point
        }),
      },
    }))
  }

  const handleDeletePoint = (indexInTable) => {
    setQueryToUpdate((oldQuery) => ({
      ...oldQuery,
      settings: {
        ...oldQuery.settings,
        points: oldQuery.settings.points.filter((_point, index) => index !== indexInTable),
      },
    }))
    Object.keys(errors).forEach((errorKey) => {
      if (errorKey.split('.')[1] === String(indexInTable)) delete errors[errorKey]
    })
  }

  const handleDeleteAllPoint = () => {
    setQueryToUpdate({ ...queryToUpdate, settings: { ...queryToUpdate.settings, points: [] } })
  }

  const handleImportPoints = async (file) => {
    try {
      const text = await utils.readFileContent(file)
      utils
        .parseCSV(text)
        .then((newPoints) => {
          setQueryToUpdate({
            ...queryToUpdate,
            settings: { ...queryToUpdate.settings, points: [...queryToUpdate.settings.points, ...newPoints] },
          })
        })
        .catch((error) => {
          console.error(error)
          setAlert({ text: error.message, type: 'danger' })
        })
    } catch (error) {
      console.error(error)
      setAlert({ text: error.message, type: 'danger' })
    }
  }

  const checkIfQueryHasChanged = () => JSON.stringify(queryToUpdate.settings) === JSON.stringify(query.settings)
        || queryToUpdate.startTime !== query.startTime || queryToUpdate.endTime !== query.endTime
        || queryToUpdate.filePattern !== query.filePattern
        || queryToUpdate.compress !== query.compress

  /**
     * Returns the text color for each status
     * @param {string} status The status of the current history query
     * @returns {string} The color represented by the status
     */
  const statusColor = (status) => {
    switch (status) {
      case 'pending':
        return 'text-warning mx-1'
      case 'exporting':
        return 'text-info mx-1'
      case 'importing':
        return 'text-primary mx-1'
      case 'finished':
        return 'text-success mx-1'
      default:
        return 'text-primary mx-1'
    }
  }

  return (
    <>
      <div className="d-flex align-items-center w-100 oi-sub-nav mb-2">
        <h6 className="text-muted d-flex align-items-center pl-3 pt-1 ml-2">
          <Button
            id="oi-navigate"
            outline
            onClick={() => {
              navigate(-1)
            }}
            className="util-button"
          >
            <FaArrowLeft className="oi-back-icon mr-2" />
          </Button>
          {`| ${name}`}
        </h6>
      </div>
      <Container fluid>
        <Form className="m-2">
          <OIbTitle label="Handlers" />
          <Row className="mb-2">
            <Col md={2}>
              <ConnectorButton
                connectorName={dataSource?.name ?? ''}
                connectorUrl={`/south/${queryToUpdate.southId}`}
              />
              <FaArrowRight className="me-2" />
              <ConnectorButton
                connectorName={application?.name ?? ''}
                connectorUrl={`/north/${queryToUpdate.northId}`}
              />
            </Col>
          </Row>
          <OIbTitle label="General settings" />
          <Row>
            <Col md={2}>
              <OIbCheckBox
                name="enabled"
                label="Enabled"
                defaultValue={false}
                value={!!queryToUpdate.enabled}
                help={<div>Enable this history query</div>}
                onChange={handleEnabled}
                switchButton
              />
            </Col>
            {queryToUpdate.enabled && progressStatus.status !== 'finished' ? (
              <Col md={2}>
                {paused ? (
                  <>
                    <FaPlayCircle
                      className="oi-icon-breadcrumb"
                      size={15}
                      onClick={(e) => {
                        e.preventDefault()
                        handlePause()
                      }}
                    />

                    <Label className="status-text-breadcrumb text-warning">
                      Paused
                    </Label>
                  </>
                ) : (
                  <>
                    <FaPauseCircle
                      className="oi-icon-breadcrumb"
                      size={15}
                      onClick={(e) => {
                        e.preventDefault()
                        handlePause()
                      }}
                    />
                    <Label className="status-text-breadcrumb text-success">
                      Ongoing
                    </Label>
                  </>
                )}
              </Col>
            ) : null}
            <Col md={2}>
              <Button
                disabled={checkIfQueryHasChanged()}
                color="primary"
                onClick={() => handleUpdateHistoryQuery()}
              >
                Save
              </Button>
            </Col>
          </Row>
          <OIbTitle label="Progress status" />
          <Row className="mb-2">
            <Col md={3}>
              <div>
                <span><b>Status:</b></span>
                <span className={statusColor(progressStatus.status)}>{progressStatus.status}</span>
              </div>
            </Col>
            {progressStatus.currentTime ? (
              <Col md={3}>
                <div>
                  <span><b>Current time:</b></span>
                  <span className="mx-1">{progressStatus.currentTime}</span>
                </div>
              </Col>
            ) : null}
            {progressStatus.progress ? (
              <Col md={6}>
                <div>
                  <span><b>Export progress:</b></span>
                  <span className="mx-1">
                    {progressStatus.progress}
                    %
                  </span>
                  {progressStatus.scanGroup && (
                    <span className="mx-1">
                      (for scan group
                      {progressStatus.scanGroup}
                      )
                    </span>
                  )}
                </div>
              </Col>
            ) : null}
          </Row>
          <OIbTitle label="Settings" />
          <Row>
            <Col md={2}>
              <OIbDate
                name="startTime"
                value={queryToUpdate.startTime}
                label="From date"
                onChange={onChange}
              />
            </Col>
            <Col md={2}>
              <OIbDate
                name="endTime"
                value={queryToUpdate.endTime}
                label="To date"
                onChange={onChange}
              />
            </Col>
            <Col md={2}>
              <OIbInteger
                name="maxReadInterval"
                label="Max read interval (s)"
                value={queryToUpdate.settings.maxReadInterval}
                defaultValue={3600}
                onChange={onMaxReadIntervalChange}
                help={<div>Put 0 to not split the query</div>}
              />
            </Col>
            <Col md={2}>
              <OIbInteger
                name="readIntervalDelay"
                label="Read interval delay (ms)"
                value={queryToUpdate.settings.readIntervalDelay}
                defaultValue={200}
                onChange={onReadIntervalDelayChange}
              />
            </Col>
          </Row>
          <Row>
            <Col md={2}>
              <OIbText
                label="File pattern"
                name="filePattern"
                value={queryToUpdate.filePattern}
                defaultValue="./@ConnectorName-@CurrentDate-@QueryPart.csv"
                onChange={onChange}
              />
            </Col>
            <Col md={2}>
              <OIbCheckBox
                name="compress"
                label="Compress"
                defaultValue={false}
                value={!!queryToUpdate.compress}
                onChange={onChange}
              />
            </Col>
          </Row>
          <Row>
            <Col md={8}>
              {queryToUpdate.settings.points ? (
                <PointsSection
                  query={queryToUpdate}
                  handleAddPoint={handleAddPoint}
                  handleChange={handleChangePoint}
                  handleDeletePoint={handleDeletePoint}
                  handleDeleteAllPoint={handleDeleteAllPoint}
                  handleImportPoints={handleImportPoints}
                />
              ) : (
                <>
                  <OIbTitle label="Request" />
                  <OIbTextArea
                    label="Query"
                    name="query"
                    contentType="sql"
                    value={queryToUpdate.settings?.query}
                    onChange={handleAddQuery}
                    help={
                      (
                        <div>
                          Available variables: @StartTime, @EndTime.
                          Be sure to use both variables to split the query with the Max read
                          interval field.
                        </div>
                      )
                    }
                  />
                </>
              )}
            </Col>
          </Row>
        </Form>
      </Container>
    </>
  )
}

HistoryQueryForm.propTypes = { query: PropTypes.object.isRequired }

export default HistoryQueryForm
