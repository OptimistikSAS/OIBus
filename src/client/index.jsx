import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'
import { Container } from 'reactstrap'

import './style/main.less'

import TopHeader from './TopHeader.jsx'
import Activation from './Activation/Activation.jsx'
import NotFound from './NotFound.jsx'
import South from './South/South.jsx'
import North from './North/North.jsx'
import Engine from './Engine/Engine.jsx'
import ConfigureApi from './North/ConfigureApi.jsx'
import ConfigureProtocol from './South/ConfigureProtocol.jsx'
import ConfigurePoints from './South/ConfigurePoints.jsx'
import Logs from './Logs/Logs.jsx'
import Health from './Health/Health.jsx'
import AlertContainer from './components/AlertContainer.jsx'
import { AlertProvider } from './context/AlertContext.jsx'
import { ConfigProvider } from './context/configContext.jsx'

const Main = () => (
  <Router>
    <>
      <ConfigProvider>
        <AlertProvider>
          <TopHeader />
          <Container className="oi-container-with-top-nav" fluid>
            <AlertContainer />
            <Switch>
              <Route exact path="/" component={Health} />
              <Route exact path="/engine" component={Engine} />
              <Route exact path="/south" component={South} />
              <Route exact path="/south/:dataSourceId" component={ConfigureProtocol} />
              <Route exact path="/south/:dataSourceId/points" component={ConfigurePoints} />
              <Route exact path="/north" component={North} />
              <Route exact path="/north/:applicationId" component={ConfigureApi} />
              <Route exact path="/log" component={Logs} />
              <Route exact path="/activation" component={Activation} />
              <Route component={NotFound} />
            </Switch>
          </Container>
        </AlertProvider>
      </ConfigProvider>
    </>
  </Router>
)

ReactDOM.render(<Main />, document.getElementById('root'))
