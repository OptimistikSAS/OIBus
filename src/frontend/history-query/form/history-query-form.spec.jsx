/**
 * @jest-environment jsdom
 */
import React from 'react'
import { act } from 'react-dom/test-utils'

import * as ReactDOMClient from 'react-dom/client'
import { testHistoryConfig } from '../../../../tests/test-config.js'
import HistoryQueryForm from './history-query-form.jsx'

React.useContext = jest.fn().mockReturnValue({ newConfig: testHistoryConfig })
jest.mock('react-router-dom', () => (
  { useNavigate: jest.fn().mockReturnValue({}) }
))
global.EventSource = class {
  constructor() {
    this.close = () => {}
  }
}
const mockMath = Object.create(global.Math)
mockMath.random = () => 1
global.Math = mockMath

let container
let root
// eslint-disable-next-line no-undef
globalThis.IS_REACT_ACT_ENVIRONMENT = true
beforeEach(() => {
  container = document.createElement('div')
  root = ReactDOMClient.createRoot(container)
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
  root = null
})

describe('HistoryQueryForm', () => {
  testHistoryConfig.forEach((historyQuery) => {
    test(`check HistoryQueryForm with data: ${historyQuery.name}`, () => {
      act(() => {
        root.render(
          <HistoryQueryForm
            query={historyQuery}
          />,
        )
      })
      expect(container).toMatchSnapshot()
    })
  })

  test('check HistoryQueryForm with empty North', () => {
    act(() => {
      root.render(
        <HistoryQueryForm
          query={{
            id: 'unique-id',
            name: 'north -> south',
            enabled: false,
            status: 'pending',
            southId: 'southId',
            northId: 'northId',
            query: '',
            order: 1,
            settings: {},
          }}
        />,
      )
    })
    expect(container).toMatchSnapshot()
  })
})
