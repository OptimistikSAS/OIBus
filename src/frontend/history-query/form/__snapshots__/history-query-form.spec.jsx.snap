// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryQueryForm check HistoryQueryForm with data: MSSQL South -> Console 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav mb-2"
  >
    <h6
      class="text-muted d-flex align-items-center pl-3 pt-1 ml-2"
    >
      <button
        class="util-button btn btn-outline-secondary"
        id="oi-navigate"
        type="button"
      >
        <svg
          class="oi-back-icon mr-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          />
        </svg>
      </button>
      | MSSQL South -&gt; Console
    </h6>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          Handlers
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-2"
        >
          <button
            class="inline-button autosize oi-status-button btn btn-outline-primary btn-sm"
            type="button"
          />
          <svg
            class="me-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
            />
          </svg>
          <button
            class="inline-button autosize oi-status-button btn btn-outline-primary btn-sm"
            type="button"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          General settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Enabled
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="enabled"
              name="enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this history query
              </div>
            </small>
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Save
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          Progress status
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-3"
        >
          <div>
            <span>
              <b>
                Status:
              </b>
            </span>
            <span
              class="text-warning mx-1"
            >
              pending
            </span>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          Settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <div
              class="align-items-sm-baseline row"
            >
              <div
                class="col-sm-auto col"
              >
                <label
                  class="form-label"
                  for="startTimeDate"
                >
                  From date
                </label>
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="startTimeDate"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="date"
                  value="2021-10-15"
                />
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="startTimeTime"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="time"
                  value="14:18"
                />
              </div>
              <div
                class="invalid-feedback"
              />
            </div>
            <div
              class="row"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <div
              class="align-items-sm-baseline row"
            >
              <div
                class="col-sm-auto col"
              >
                <label
                  class="form-label"
                  for="endTimeDate"
                >
                  To date
                </label>
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="endTimeDate"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="date"
                  value="2021-10-16"
                />
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="endTimeTime"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="time"
                  value="14:18"
                />
              </div>
              <div
                class="invalid-feedback"
              />
            </div>
            <div
              class="row"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="maxReadInterval"
            >
              Max read interval (s)
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="maxReadInterval"
              name="maxReadInterval"
              type="integer"
              value=""
            />
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Put 0 to not split the query
              </div>
            </small>
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="readIntervalDelay"
            >
              Read interval delay (ms)
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="readIntervalDelay"
              name="readIntervalDelay"
              type="integer"
              value="200"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="filePattern"
            >
              File pattern
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="filePattern"
              name="filePattern"
              type="text"
              value="history1-@CurrentDate.csv"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Compress
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="compress"
              name="compress"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-8"
        >
          <div
            class="spinner-container"
          >
            <div
              class="spinner-grow text-primary"
              role="status"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
            ...loading points from OIBus server...
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`HistoryQueryForm check HistoryQueryForm with data: MSSQL South -> OIAnalytics North 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav mb-2"
  >
    <h6
      class="text-muted d-flex align-items-center pl-3 pt-1 ml-2"
    >
      <button
        class="util-button btn btn-outline-secondary"
        id="oi-navigate"
        type="button"
      >
        <svg
          class="oi-back-icon mr-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          />
        </svg>
      </button>
      | MSSQL South -&gt; OIAnalytics North
    </h6>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          Handlers
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-2"
        >
          <button
            class="inline-button autosize oi-status-button btn btn-outline-primary btn-sm"
            type="button"
          />
          <svg
            class="me-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
            />
          </svg>
          <button
            class="inline-button autosize oi-status-button btn btn-outline-primary btn-sm"
            type="button"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          General settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Enabled
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="enabled"
              name="enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this history query
              </div>
            </small>
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Save
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          Progress status
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-3"
        >
          <div>
            <span>
              <b>
                Status:
              </b>
            </span>
            <span
              class="text-info mx-1"
            >
              exporting
            </span>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          Settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <div
              class="align-items-sm-baseline row"
            >
              <div
                class="col-sm-auto col"
              >
                <label
                  class="form-label"
                  for="startTimeDate"
                >
                  From date
                </label>
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="startTimeDate"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="date"
                  value="2021-10-15"
                />
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="startTimeTime"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="time"
                  value="14:18"
                />
              </div>
              <div
                class="invalid-feedback"
              />
            </div>
            <div
              class="row"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <div
              class="align-items-sm-baseline row"
            >
              <div
                class="col-sm-auto col"
              >
                <label
                  class="form-label"
                  for="endTimeDate"
                >
                  To date
                </label>
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="endTimeDate"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="date"
                  value="2021-10-16"
                />
              </div>
              <div
                class="col-sm-auto col"
              >
                <input
                  class="oi-form-input form-control"
                  id="endTimeTime"
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required=""
                  type="time"
                  value="14:18"
                />
              </div>
              <div
                class="invalid-feedback"
              />
            </div>
            <div
              class="row"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="maxReadInterval"
            >
              Max read interval (s)
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="maxReadInterval"
              name="maxReadInterval"
              type="integer"
              value=""
            />
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Put 0 to not split the query
              </div>
            </small>
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="readIntervalDelay"
            >
              Read interval delay (ms)
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="readIntervalDelay"
              name="readIntervalDelay"
              type="integer"
              value="200"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="filePattern"
            >
              File pattern
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="filePattern"
              name="filePattern"
              type="text"
              value="history1-@CurrentDate.csv"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Compress
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="compress"
              name="compress"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-8"
        >
          <div
            class="spinner-container"
          >
            <div
              class="spinner-grow text-primary"
              role="status"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
            ...loading points from OIBus server...
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
